client/iotcore/client.go

```diff
@@ -0,0 +1,315 @@
+	svc := iot.New(sess)
+
+	awsDevice, err := c.GetDevice(ctx, sess, deviceID)
+	if err == nil && awsDevice != nil {
+		cert, err := svc.DescribeCertificateWithContext(ctx, &iot.DescribeCertificateInput{
+			CertificateId: aws.String(awsDevice.CertificateID),
+		})
+		if err == nil {
+			newStatus := iot.CertificateStatusInactive
+			awsDevice.Status = StatusDisabled
+			if device.Status == StatusEnabled {
+				newStatus = iot.CertificateStatusActive
+				awsDevice.Status = StatusEnabled
+			}
+
+			if *cert.CertificateDescription.Status != newStatus {
+				paramsUpdateCertificate := &iot.UpdateCertificateInput{
+					CertificateId: aws.String(awsDevice.CertificateID),
+					NewStatus:     aws.String(newStatus),
+				}
+				_, err = svc.UpdateCertificateWithContext(ctx, paramsUpdateCertificate)
+			}
+		}
+
+		return awsDevice, err
+	} else if err == ErrDeviceNotFound {
+		err = nil
+	}
+
+	var resp *iot.CreateThingOutput
+	if err == nil {
+		resp, err = svc.CreateThingWithContext(ctx,
+			&iot.CreateThingInput{
+				ThingName: aws.String(deviceID),
+			})
+	}
+
+	var respPolicy *iot.CreatePolicyOutput
+	if err == nil {
+		respPolicy, err = svc.CreatePolicyWithContext(ctx,
+			&iot.CreatePolicyInput{
+				PolicyDocument: aws.String(policy),
+				PolicyName:     aws.String(policyName(deviceID)),
+			})
+	}
+	var privKey *ecdsa.PrivateKey
+	if err == nil {
+		privKey, err = newPrivateKey()
+	}
+
+	var csr []byte
+	if err == nil {
+		csr, err = newCSR(deviceID, privKey)
+	}
+
+	var respCert *iot.CreateCertificateFromCsrOutput
+	if err == nil {
+		respCert, err = svc.CreateCertificateFromCsrWithContext(ctx,
+			&iot.CreateCertificateFromCsrInput{
+				CertificateSigningRequest: aws.String(string(csr)),
+				SetAsActive:               aws.Bool(device.Status == StatusEnabled),
+			})
+	}
+
+	if err == nil {
+		_, err = svc.AttachPolicyWithContext(ctx,
+			&iot.AttachPolicyInput{
+				PolicyName: respPolicy.PolicyName,
+				Target:     respCert.CertificateArn,
+			})
+	}
+
+	if err == nil {
+		_, err = svc.AttachThingPrincipalWithContext(ctx,
+			&iot.AttachThingPrincipalInput{
+				Principal: respCert.CertificateArn,
+				ThingName: aws.String(deviceID),
+			})
+	}
+
+	var deviceResp *Device
+	if err == nil {
+		deviceResp = &Device{
+			ID:          *resp.ThingId,
+			Name:        *resp.ThingName,
+			Status:      device.Status,
+			PrivateKey:  string(privateKeyToPem(privKey)),
+			Certificate: *respCert.CertificatePem,
+		}
+	}
+	return deviceResp, err
+}
+
+func (c *client) DeleteDevice(ctx context.Context, sess *session.Session, deviceID string) error {
+	svc := iot.New(sess)
+
+	respDescribe, err := svc.DescribeThingWithContext(ctx,
+		&iot.DescribeThingInput{
+			ThingName: aws.String(deviceID),
+		})
+
+	var respListThingPrincipals *iot.ListThingPrincipalsOutput
+	if err == nil {
+		respListThingPrincipals, err = svc.ListThingPrincipalsWithContext(ctx,
+			&iot.ListThingPrincipalsInput{
+				ThingName: aws.String(deviceID),
+			})
```

Comment:
nit: exact the same is repeated twice, perhaps a helper function?

Reply:
Don't know, this is going to increase the code complexity for nothing (at the end, you replace one SDK func call with a helper func call).